File: packages/activity_recognition_flutter/android/src/main/java/dk/cachet/activity_recognition_flutter/ActivityRecognizedService.java
package dk.cachet.activity_recognition_flutter;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.core.app.JobIntentService;

import com.google.android.gms.location.ActivityRecognitionResult;
import com.google.android.gms.location.DetectedActivity;

import java.util.List;

public class ActivityRecognizedService extends JobIntentService {

    static void enqueueWork(Context context, Intent work) {
        enqueueWork(context, ActivityRecognizedService.class, 1, work);
    }

    @Override
    public int onStartCommand(@Nullable Intent intent, int flags, int startId) {
        return super.onStartCommand(intent, flags, startId);
    }

    // This method is called when service starts instead of onHandleIntent
    protected void onHandleWork(@Nullable Intent intent) {
        onHandleIntent(intent);
    }

    // remove override and make onHandleIntent private.
    private void onHandleIntent(@Nullable Intent intent) {
        ActivityRecognitionResult result = ActivityRecognitionResult.extractResult(intent);
        List<DetectedActivity> activities = result.getProbableActivities();

        DetectedActivity mostLikely = activities.get(0);

        for (DetectedActivity a : activities) {
            if (a.getConfidence() > mostLikely.getConfidence()) {
                mostLikely = a;
            }
        }

        String type = getActivityString(mostLikely.getType());
        int confidence = mostLikely.getConfidence();

        String data = type + "," + confidence;

        Log.d("onHandleIntent", data);

        // Same preferences as in ActivityRecognitionFlutterPlugin.java
        SharedPreferences preferences =
                getApplicationContext().getSharedPreferences(
                        ActivityRecognitionFlutterPlugin.ACTIVITY_RECOGNITION, MODE_PRIVATE);

        preferences.edit().clear()
                .putString(
                        ActivityRecognitionFlutterPlugin.DETECTED_ACTIVITY,
                        data
                )
                .apply();
    }

    public static String getActivityString(int type) {
        if (type == DetectedActivity.IN_VEHICLE) return "IN_VEHICLE";
        if (type == DetectedActivity.ON_BICYCLE) return "ON_BICYCLE";
        if (type == DetectedActivity.ON_FOOT) return "ON_FOOT";
        if (type == DetectedActivity.RUNNING) return "RUNNING";
        if (type == DetectedActivity.STILL) return "STILL";
        if (type == DetectedActivity.TILTING) return "TILTING";
        if (type == DetectedActivity.WALKING) return "WALKING";

        // Default case
        return "UNKNOWN";
    }
}

File: packages/activity_recognition_flutter/android/src/main/kotlin/dk/cachet/activity_recognition_flutter/activity/ActivityRecognizedService.kt
/*
 * Copyright (c) 2018. Daniel Morawetz
 * Licensed under Apache License v2.0
 */

package dk.cachet.activity_recognition_flutter.activity

import android.app.IntentService
import android.content.Context
import android.content.Intent
import android.os.Build
import android.preference.PreferenceManager
import android.util.Log
import dk.cachet.activity_recognition_flutter.Codec
import dk.cachet.activity_recognition_flutter.Constants
import com.google.android.gms.location.ActivityRecognitionResult

class ActivityRecognizedService(name: String = "ActivityRecognizedService") : IntentService(name) {

    val TAG = "ActivityRecognizedServi"

    override fun onHandleIntent(intent: Intent) {
        Log.d(TAG, "received activity update!")
        val result = ActivityRecognitionResult.extractResult(intent)

        // Get the list of the probable activities associated with the current state of the
        // device. Each activity is associated with a confidence level, which is an int between
        // 0 and 100.
        val detectedActivities = result.probableActivities as ArrayList


        val mostProbableActivity = detectedActivities.maxBy { it.confidence }

        val preferences =
                applicationContext.getSharedPreferences("activity_recognition", MODE_PRIVATE)
        preferences.edit().clear()
                .putString(
                        Constants.KEY_DETECTED_ACTIVITIES,
                        Codec.encodeResult(mostProbableActivity!!)
                )
                .apply()
    }
}